File: apps/ios/BrickBack/Features/Catalog/SetDetailScreen.swift
import SwiftUI
import BrickBackKit

/// Set detail (`.setDetail`). Renders catalog metadata and the primary "Start sorting" action,
/// which snapshots the set into local GRDB and opens it. Port of `set_detail_screen.dart`.
/// Adding a set is unlimited (no free-tier cap) — premium gates only party mode + cloud sync.
struct SetDetailScreen: View {
    @Environment(AppEnvironment.self) private var env
    @Environment(\.activeRouter) private var activeRouter
    let itemId: Int

    @State private var state: LoadState<SetDetail> = .loading
    /// Unique (part, colour) count, loaded lazily so metadata renders immediately ("…" until ready).
    @State private var uniqueParts = "…"

    private var router: Router { activeRouter ?? env.homeRouter }

    /// The native bar's inline title: the set's own name once loaded (the big content heading in a
    /// compact form the bar keeps as you scroll), falling back to the generic label while it loads.
    private var navTitle: String {
        if case .loaded(let detail) = state { return detail.set.name }
        return L.setHeader
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 0) {
            switch state {
            case .idle, .loading:
                ProgressView().tint(AppColors.primary).frame(maxWidth: .infinity, maxHeight: .infinity)
            case .failed(let message):
                EmptyState(title: L.setCouldntLoad, message: message, icon: "exclamationmark.triangle")
            case .loaded(let detail):
                Detail(detail: detail, uniqueParts: uniqueParts)
            }
        }
        .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .topLeading)
        .background(AppColors.canvas)
        // Only shown when the native bar is present (the search tab); harmless where it's hidden.
        .navigationTitle(navTitle)
        .navigationBarTitleDisplayMode(.inline)
        .task(id: itemId) {
            state = .loading
            do {
                let d = try await env.services.setDetail(itemId)
                if !Task.isCancelled { state = .loaded(d) }
            } catch {
                if !Task.isCancelled { state = .failed("\(error)") }
            }
        }
        .task(id: itemId) {
            // Load the unique-part count alongside the metadata (the two providers in Flutter).
            if let parts = try? await env.services.catalog.expandSetParts(itemId), !Task.isCancelled {
                uniqueParts = "\(parts.count)"
            }
        }
    }
}

private struct Detail: View {
    @Environment(AppEnvironment.self) private var env
    @Environment(\.activeRouter) private var activeRouter
    let detail: SetDetail
    let uniqueParts: String

    private var router: Router { activeRouter ?? env.homeRouter }

    // MARK: Lifecycle + value

    /// The coloured status pill (nil when the catalog has no stage for this set).
    private var lifecycleBadge: (text: String, color: Color)? {
        guard let lifecycle = detail.set.lifecycle else { return nil }
        switch lifecycle {
        case .upcoming:     return (L.lifecycleUpcoming, AppColors.info)
        case .available:    return (L.lifecycleAvailable, AppColors.success)
        case .retiringSoon: return (L.lifecycleRetiringSoon, AppColors.warning)
        case .retired:      return (L.lifecycleRetired, AppColors.inkSoft)
        }
    }

    /// Show the badge only when the Availability card doesn't already state the same thing: the
    /// dated retired/retiring/upcoming rows make the pill redundant, so it survives only for
    /// currently-available sets (which have no such row) or when there are no dates to show.
    private var showLifecycleBadge: Bool {
        guard let lifecycle = detail.set.lifecycle else { return false }
        switch lifecycle {
        case .retired, .retiringSoon: return retirementRow == nil
        case .upcoming:               return releaseRow == nil
        case .available:              return true
        }
    }

    /// Release-date row — labelled by tense (upcoming sets haven't released yet).
    private var releaseRow: (label: String, value: String)? {
        guard let date = detail.set.launchDate else { return nil }
        let label = detail.set.lifecycle == .upcoming ? L.dateReleases : L.dateReleased
        return (label, monthYear(date))
    }

    /// Retirement-date row — the exact exit date once retired, the estimate while retiring soon.
    private var retirementRow: (label: String, value: String)? {
        guard let lifecycle = detail.set.lifecycle else { return nil }
        switch lifecycle {
        case .retired:
            guard let date = detail.set.exitDate else { return nil }
            return (L.dateRetired, monthYear(date))
        case .retiringSoon:
            guard let date = detail.set.retiringSoonDate ?? detail.set.exitDate else { return nil }
            return (L.dateRetiring, monthYear(date))
        default:
            return nil
        }
    }

    @ViewBuilder private var availabilitySection: some View {
        if releaseRow != nil || retirementRow != nil {
            Spacer().frame(height: AppSpacing.s20)
            sectionHeader(L.availabilityTitle)
            Spacer().frame(height: AppSpacing.s8)
            AppCard {
                VStack(alignment: .leading, spacing: AppSpacing.s12) {
                    if let r = releaseRow { InfoRow(label: r.label, value: r.value) }
                    if let r = retirementRow { InfoRow(label: r.label, value: r.value) }
                }
            }
        }
    }

    @ViewBuilder private var valueSection: some View {
        if let price = detail.price, price.hasAny {
            Spacer().frame(height: AppSpacing.s20)
            sectionHeader(L.valueTitle)
            Spacer().frame(height: AppSpacing.s8)
            AppCard {
                VStack(alignment: .leading, spacing: AppSpacing.s12) {
                    if let n = price.new {
                        InfoRow(label: L.valueNew, value: money(n, price.currency), valueColor: AppColors.success)
                    }
                    if let u = price.used {
                        InfoRow(label: L.valueUsed, value: money(u, price.currency), valueColor: AppColors.warning)
                    }
                }
            }
        }
    }

    private func sectionHeader(_ text: String) -> some View {
        Text(text).font(AppText.h2).foregroundStyle(AppColors.ink)
    }

    /// Localised month + year (e.g. "June 2013"). Community dates are month-precision at best, so
    /// we deliberately drop the day. Formatted in UTC to match how the "yyyy-MM-dd" value parsed.
    private func monthYear(_ date: Date) -> String {
        let f = DateFormatter()
        f.locale = I18n.locale
        f.timeZone = TimeZone(identifier: "UTC")
        f.setLocalizedDateFormatFromTemplate("yMMMM")
        return f.string(from: date)
    }

    private func money(_ value: Double, _ currency: String) -> String {
        let f = NumberFormatter()
        f.numberStyle = .currency
        f.currencyCode = currency
        f.locale = I18n.locale
        f.maximumFractionDigits = 2
        return f.string(from: NSNumber(value: value)) ?? String(format: "%.2f", value)
    }

    var body: some View {
        ScrollView {
            VStack(alignment: .leading, spacing: 0) {
                SetThumb(imageUrl: detail.set.imageUrl, size: 200, radius: AppRadius.lg)
                    .frame(maxWidth: .infinity)
                Spacer().frame(height: AppSpacing.s16)
                if let theme = detail.themeName, !theme.isEmpty {
                    Text(theme).font(AppText.title).foregroundStyle(AppColors.inkSoft)
                    Spacer().frame(height: AppSpacing.s8)
                }
                WrapLayout(spacing: AppSpacing.s8, lineSpacing: AppSpacing.s8) {
                    if !detail.set.setNum.isEmpty { AppBadge(detail.set.setNum) }
                    // Year only when the Availability card won't already show a release date (else
                    // it just repeats it); keeps the year visible for older, undated sets.
                    if detail.set.launchDate == nil, detail.set.year != 0 {
                        AppBadge(String(detail.set.year))
                    }
                    AppBadge(partsCountLabel(detail.set.numParts))
                    // Status rides the same row (it's just one more chip) and wraps only if needed.
                    if showLifecycleBadge, let badge = lifecycleBadge {
                        AppBadge(badge.text, color: badge.color)
                    }
                }
                Spacer().frame(height: AppSpacing.s16)
                HStack(spacing: AppSpacing.s12) {
                    StatCard(label: L.uniqueParts, value: uniqueParts) {
                        router.push(.setParts(detail.set.itemId))
                    }
                    StatCard(label: L.minifigs, value: "\(detail.minifigCount)") {
                        router.push(.setMinifigs(detail.set.itemId))
                    }
                }
                availabilitySection
                valueSection
                Spacer().frame(height: AppSpacing.s20)
                StartSortingButton(itemId: detail.set.itemId)
                Spacer().frame(height: AppSpacing.s12)
                Text(L.startSortingHint)
                    .font(AppText.caption).foregroundStyle(AppColors.inkSoft)
                    .multilineTextAlignment(.center)
                    .frame(maxWidth: .infinity)
            }
            .padding(.horizontal, AppSpacing.screen)
            .padding(.bottom, AppSpacing.s24)
        }
    }
}

private struct StatCard: View {
    let label: String
    let value: String
    let onTap: () -> Void

    var body: some View {
        AppCard(padding: AppSpacing.s12, onTap: onTap) {
            VStack(spacing: 2) {
                Text(value).font(AppText.h1).foregroundStyle(AppColors.ink)
                HStack(spacing: 2) {
                    Text(label).font(AppText.caption).foregroundStyle(AppColors.inkSoft).lineLimit(1)
                    Image(systemName: "chevron.right").font(.system(size: 11)).foregroundStyle(AppColors.muted)
                }
            }
            .frame(maxWidth: .infinity)
        }
    }
}

/// A label→value line inside an info card (availability dates, market value). The value carries
/// the emphasis (and, for prices, the new/used colour); the label stays quiet.
private struct InfoRow: View {
    let label: String
    let value: String
    var valueColor: Color = AppColors.ink

    var body: some View {
        HStack(spacing: AppSpacing.s12) {
            Text(label).font(AppText.body).foregroundStyle(AppColors.inkSoft)
            Spacer(minLength: 0)
            Text(value).font(AppText.title).foregroundStyle(valueColor)
        }
    }
}

private struct StartSortingButton: View {
    @Environment(AppEnvironment.self) private var env
    @Environment(\.activeRouter) private var activeRouter
    let itemId: Int
    @State private var loading = false
    @State private var errorMessage: String?

    private var router: Router { activeRouter ?? env.homeRouter }

    var body: some View {
        AppButton(L.startSorting, icon: "checklist", loading: loading, expand: true) {
            guard !loading else { return }
            start()
        }
        .alert(L.couldntAddSet, isPresented: Binding(
            get: { errorMessage != nil },
            set: { if !$0 { errorMessage = nil } }
        )) {
            Button(L.ok, role: .cancel) {}
        } message: {
            Text(errorMessage ?? "")
        }
    }

    private func start() {
        loading = true
        Task {
            do {
                let id = try await env.services.rebuild.addSet(itemId)
                // Get the new rebuild to the cloud promptly once premium sync is live (no-op now).
                env.sync.nudge()
                // S9: eagerly cache this set's images for offline while we're still online.
                Task { await env.services.offlineImages.ensureCached(id) }
                loading = false
                // Starting the build ends the "add set" flow: clear the stack we came in on and
                // open counting over on Rebuilds.
                env.openRebuild(id, clearing: router)
            } catch {
                loading = false
                errorMessage = "\(error)"
            }
        }
    }
}

File: apps/mobile/lib/features/catalog/set_detail_screen.dart
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'package:go_router/go_router.dart';
import 'package:intl/intl.dart';

import '../../core/sync/sync_service.dart';
import '../../l10n/l10n.dart';
import '../../theme/app_theme.dart';
import '../../widgets/primitives.dart';
import '../../widgets/readable_column.dart';
import '../rebuild/rebuild_repository.dart';
import 'catalog_models.dart';
import 'catalog_repository.dart';

/// Set detail (`/set/:id`). A collapsing blue hero header (back + owned-check +
/// the set box) over the metadata + the primary "Start Building" action, which
/// snapshots the set into local Drift and opens it.
class SetDetailScreen extends ConsumerWidget {
  const SetDetailScreen({super.key, required this.itemId});

  final int itemId;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final c = BrickColors.of(context);
    final detail = ref.watch(setDetailProvider(itemId));

    return AnnotatedRegion<SystemUiOverlayStyle>(
      value: SystemUiOverlayStyle.light,
      child: Scaffold(
        backgroundColor: c.canvas,
        body: detail.when(
          loading: () => _LoadingOrError(
            itemId: itemId,
            child: Center(child: CircularProgressIndicator(color: c.primary)),
          ),
          error: (e, _) => _LoadingOrError(
            itemId: itemId,
            child: EmptyState(
              icon: Icons.error_outline,
              title: context.l10n.setCouldntLoad,
              message: '$e',
            ),
          ),
          data: (d) => _Detail(detail: d),
        ),
      ),
    );
  }
}

/// Loading / error still get the blue header bar with a working back button.
class _LoadingOrError extends StatelessWidget {
  const _LoadingOrError({required this.itemId, required this.child});
  final int itemId;
  final Widget child;
  @override
  Widget build(BuildContext context) {
    final c = BrickColors.of(context);
    final topPad = MediaQuery.paddingOf(context).top;
    return Column(
      children: [
        Container(
          height: topPad + 72,
          decoration: BoxDecoration(
            gradient: LinearGradient(
              begin: Alignment.topCenter,
              end: Alignment.bottomCenter,
              colors: [c.brandDeep, c.brand],
            ),
            borderRadius: const BorderRadius.vertical(bottom: Radius.circular(AppRadius.xl)),
          ),
          padding: EdgeInsets.only(top: topPad + 8, left: AppSpacing.screen),
          child: Align(
            alignment: Alignment.centerLeft,
            child: SquircleButton(
              icon: Icons.arrow_back_rounded,
              fill: c.brand,
              edge: c.brandEdge,
              fg: Colors.white,
              onTap: () => Navigator.of(context).maybePop(),
              semanticLabel: MaterialLocalizations.of(context).backButtonTooltip,
            ),
          ),
        ),
        Expanded(child: child),
      ],
    );
  }
}

class _Detail extends ConsumerWidget {
  const _Detail({required this.detail});
  final SetDetail detail;

  ({String text, Color color})? _lifecycleBadge(BuildContext context) {
    final c = BrickColors.of(context);
    switch (detail.set.lifecycle) {
      case null:
        return null;
      case SetLifecycle.upcoming:
        return (text: context.l10n.lifecycleUpcoming, color: c.info);
      case SetLifecycle.available:
        return (text: context.l10n.lifecycleAvailable, color: c.success);
      case SetLifecycle.retiringSoon:
        return (text: context.l10n.lifecycleRetiringSoon, color: c.warning);
      case SetLifecycle.retired:
        return (text: context.l10n.lifecycleRetired, color: c.danger);
    }
  }

  bool _showLifecycleBadge(BuildContext context) {
    switch (detail.set.lifecycle) {
      case null:
        return false;
      case SetLifecycle.retired:
      case SetLifecycle.retiringSoon:
        return _retirementRow(context) == null;
      case SetLifecycle.upcoming:
        return _releaseRow(context) == null;
      case SetLifecycle.available:
        return true;
    }
  }

  ({String label, String value})? _releaseRow(BuildContext context) {
    final date = detail.set.launchDate;
    if (date == null) return null;
    final label = detail.set.lifecycle == SetLifecycle.upcoming
        ? context.l10n.dateReleases
        : context.l10n.dateReleased;
    return (label: label, value: _monthYear(context, date));
  }

  ({String label, String value})? _retirementRow(BuildContext context) {
    switch (detail.set.lifecycle) {
      case SetLifecycle.retired:
        final date = detail.set.exitDate;
        if (date == null) return null;
        return (label: context.l10n.dateRetired, value: _monthYear(context, date));
      case SetLifecycle.retiringSoon:
        final date = detail.set.retiringSoonDate ?? detail.set.exitDate;
        if (date == null) return null;
        return (label: context.l10n.dateRetiring, value: _monthYear(context, date));
      default:
        return null;
    }
  }

  String _monthYear(BuildContext context, DateTime date) =>
      DateFormat.yMMMM(context.l10n.localeName).format(date.toUtc());

  String _money(BuildContext context, double value, String currency) =>
      NumberFormat.simpleCurrency(locale: context.l10n.localeName, name: currency).format(value);

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final c = BrickColors.of(context);
    final set = detail.set;
    final topPad = MediaQuery.paddingOf(context).top;

    final uniqueParts = ref.watch(setPartsProvider(set.itemId)).maybeWhen(
          data: (p) => '${p.length}',
          orElse: () => '…',
        );

    final badge = _lifecycleBadge(context);
    final releaseRow = _releaseRow(context);
    final retirementRow = _retirementRow(context);
    final price = detail.price;

    return ReadableColumn(
      child: CustomScrollView(
        slivers: [
          SliverPersistentHeader(
            pinned: true,
            delegate: CollapsingHeaderDelegate(
              minExtent: topPad + 72,
              maxExtent: topPad + 340,
              builder: (context, t) => _SetHeader(
                t: t,
                topPad: topPad,
                imageUrl: set.imageUrl,
                itemId: set.itemId,
              ),
            ),
          ),
          SliverToBoxAdapter(
            child: Padding(
              padding: const EdgeInsets.fromLTRB(
                  AppSpacing.screen, AppSpacing.s20, AppSpacing.screen, AppSpacing.s40),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.stretch,
                children: [
                  _StartSortingButton(itemId: set.itemId),
                  const SizedBox(height: AppSpacing.s8),
                  Text(context.l10n.startSortingHint,
                      style: AppText.caption.copyWith(color: c.inkSoft),
                      textAlign: TextAlign.center),
                  const SizedBox(height: AppSpacing.s24),
                  if (detail.themeName != null)
                    Text(detail.themeName!,
                        style: AppText.title.copyWith(color: c.inkSoft),
                        textAlign: TextAlign.center),
                  const SizedBox(height: AppSpacing.s4),
                  Text(set.name, style: AppText.display.copyWith(color: c.ink), textAlign: TextAlign.center),
                  if (_showLifecycleBadge(context) && badge != null) ...[
                    const SizedBox(height: AppSpacing.s16),
                    Center(child: AppBadge(badge.text, color: badge.color, filled: true)),
                  ],
                  const SizedBox(height: AppSpacing.s24),
                  // Reference 4-up info row.
                  Row(
                    children: [
                      Expanded(
                          child: StatCell(
                              icon: Icons.widgets_rounded,
                              label: context.l10n.piecesLabel,
                              value: '${set.numParts}')),
                      Expanded(
                          child: StatCell(
                              icon: Icons.face_rounded,
                              label: context.l10n.minifigs,
                              value: '${detail.minifigCount}')),
                      Expanded(
                          child: StatCell(
                              icon: Icons.event_rounded,
                              label: context.l10n.yearLabel,
                              value: set.year != 0 ? '${set.year}' : '—')),
                      Expanded(
                          child: StatCell(
                              icon: Icons.tag_rounded,
                              label: context.l10n.setLabel,
                              value: set.setNum)),
                    ],
                  ),
                  const SizedBox(height: AppSpacing.s20),
                  // Tappable drill-downs.
                  Row(
                    children: [
                      Expanded(
                        child: _LinkCard(
                          label: context.l10n.uniqueParts,
                          value: uniqueParts,
                          onTap: () => context.push('/set/${set.itemId}/parts'),
                        ),
                      ),
                      const SizedBox(width: AppSpacing.s12),
                      Expanded(
                        child: _LinkCard(
                          label: context.l10n.minifigs,
                          value: '${detail.minifigCount}',
                          onTap: () => context.push('/set/${set.itemId}/minifigs'),
                        ),
                      ),
                    ],
                  ),
                  if (releaseRow != null || retirementRow != null) ...[
                    const SizedBox(height: AppSpacing.s24),
                    _SectionHeader(context.l10n.availabilityTitle),
                    const SizedBox(height: AppSpacing.s12),
                    AppCard(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.stretch,
                        children: [
                          if (releaseRow != null)
                            _InfoRow(label: releaseRow.label, value: releaseRow.value),
                          if (releaseRow != null && retirementRow != null)
                            const SizedBox(height: AppSpacing.s12),
                          if (retirementRow != null)
                            _InfoRow(label: retirementRow.label, value: retirementRow.value),
                        ],
                      ),
                    ),
                  ],
                  if (price != null && price.hasAny) ...[
                    const SizedBox(height: AppSpacing.s24),
                    _SectionHeader(context.l10n.valueTitle),
                    const SizedBox(height: AppSpacing.s12),
                    AppCard(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.stretch,
                        children: [
                          if (price.newValue != null)
                            _InfoRow(
                              label: context.l10n.valueNew,
                              value: _money(context, price.newValue!, price.currency),
                              valueColor: c.success,
                            ),
                          if (price.newValue != null && price.used != null)
                            const SizedBox(height: AppSpacing.s12),
                          if (price.used != null)
                            _InfoRow(
                              label: context.l10n.valueUsed,
                              value: _money(context, price.used!, price.currency),
                              valueColor: c.warning,
                            ),
                        ],
                      ),
                    ),
                  ],
                ],
              ),
            ),
          ),
        ],
      ),
    );
  }
}

/// The collapsing blue header — back + owned-check buttons over the fading set box.
class _SetHeader extends StatelessWidget {
  const _SetHeader({required this.t, required this.topPad, required this.imageUrl, required this.itemId});
  final double t;
  final double topPad;
  final String? imageUrl;
  final int itemId;

  @override
  Widget build(BuildContext context) {
    final c = BrickColors.of(context);
    return ClipRRect(
      borderRadius: const BorderRadius.vertical(bottom: Radius.circular(AppRadius.xl)),
      child: DecoratedBox(
        decoration: BoxDecoration(
          gradient: LinearGradient(
            begin: Alignment.topCenter,
            end: Alignment.bottomCenter,
            colors: [c.brandDeep, c.brand],
          ),
        ),
        child: Stack(
          clipBehavior: Clip.hardEdge,
          children: [
            // The fading, shrinking set box.
            Positioned.fill(
              child: Padding(
                padding: EdgeInsets.only(
                    top: topPad + 60, left: AppSpacing.s32, right: AppSpacing.s32, bottom: AppSpacing.s20),
                child: Opacity(
                  opacity: (1 - t * 1.4).clamp(0.0, 1.0),
                  child: Center(
                    child: SetThumb(
                      imageUrl: imageUrl,
                      size: 230,
                      radius: AppRadius.lg,
                      background: Colors.transparent,
                    ),
                  ),
                ),
              ),
            ),
            // Top action row.
            Positioned(
              top: topPad + 8,
              left: AppSpacing.screen,
              right: AppSpacing.screen,
              child: Row(
                children: [
                  SquircleButton(
                    icon: Icons.arrow_back_rounded,
                    fill: c.brand,
                    edge: c.brandEdge,
                    fg: Colors.white,
                    onTap: () => Navigator.of(context).maybePop(),
                    semanticLabel: MaterialLocalizations.of(context).backButtonTooltip,
                  ),
                  const Spacer(),
                  _OwnedButton(itemId: itemId),
                ],
              ),
            ),
          ],
        ),
      ),
    );
  }
}

/// The green "owned" check — reflects whether this set is already in the
/// collection; tapping adds it (creating a rebuild) if not.
class _OwnedButton extends ConsumerWidget {
  const _OwnedButton({required this.itemId});
  final int itemId;

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final c = BrickColors.of(context);
    final owned = ref.watch(rebuildListProvider).maybeWhen(
          data: (list) => list.any((r) => r.setItemId == itemId),
          orElse: () => false,
        );
    return SquircleButton(
      icon: Icons.check_rounded,
      fill: owned ? c.success : Colors.white,
      edge: owned ? c.successEdge : c.cardEdge,
      fg: owned ? Colors.white : c.muted,
      onTap: owned
          ? () => ScaffoldMessenger.of(context).showSnackBar(
                SnackBar(content: Text(context.l10n.alreadyInCollection)),
              )
          : () async {
              try {
                await ref.read(rebuildRepositoryProvider).addSet(itemId);
                ref.invalidate(rebuildListProvider);
                ref.read(syncControllerProvider).nudge();
                if (context.mounted) {
                  ScaffoldMessenger.of(context).showSnackBar(
                    SnackBar(content: Text(context.l10n.addedToCollection)),
                  );
                }
              } catch (e) {
                if (context.mounted) {
                  ScaffoldMessenger.of(context).showSnackBar(
                    SnackBar(content: Text(context.l10n.couldntAddSet('$e'))),
                  );
                }
              }
            },
      semanticLabel: owned ? context.l10n.alreadyInCollection : context.l10n.addASet,
    );
  }
}

class _SectionHeader extends StatelessWidget {
  const _SectionHeader(this.text);
  final String text;
  @override
  Widget build(BuildContext context) =>
      Text(text, style: AppText.h2.copyWith(color: BrickColors.of(context).ink));
}

class _InfoRow extends StatelessWidget {
  const _InfoRow({required this.label, required this.value, this.valueColor});
  final String label;
  final String value;
  final Color? valueColor;

  @override
  Widget build(BuildContext context) {
    final c = BrickColors.of(context);
    return Row(
      children: [
        Text(label, style: AppText.body.copyWith(color: c.inkSoft)),
        const Spacer(),
        Text(value, style: AppText.title.copyWith(color: valueColor ?? c.ink, fontWeight: FontWeight.w800)),
      ],
    );
  }
}

/// A compact tappable metric card (Unique parts / Minifigs).
class _LinkCard extends StatelessWidget {
  const _LinkCard({required this.label, required this.value, this.onTap});
  final String label;
  final String value;
  final VoidCallback? onTap;

  @override
  Widget build(BuildContext context) {
    final c = BrickColors.of(context);
    return SoftCard(
      onTap: onTap,
      padding: const EdgeInsets.symmetric(vertical: AppSpacing.s16, horizontal: AppSpacing.s16),
      child: Column(
        children: [
          Text(value, style: AppText.h1.copyWith(color: c.ink)),
          const SizedBox(height: 2),
          Row(
            mainAxisAlignment: MainAxisAlignment.center,
            children: [
              Flexible(
                child: Text(label,
                    maxLines: 1,
                    overflow: TextOverflow.ellipsis,
                    style: AppText.caption.copyWith(color: c.inkSoft)),
              ),
              Icon(Icons.chevron_right_rounded, size: 16, color: c.muted),
            ],
          ),
        ],
      ),
    );
  }
}

class _StartSortingButton extends ConsumerStatefulWidget {
  const _StartSortingButton({required this.itemId});
  final int itemId;

  @override
  ConsumerState<_StartSortingButton> createState() => _StartSortingButtonState();
}

class _StartSortingButtonState extends ConsumerState<_StartSortingButton> {
  bool _loading = false;

  Future<void> _start() async {
    setState(() => _loading = true);
    try {
      final id = await ref.read(rebuildRepositoryProvider).addSet(widget.itemId);
      if (!mounted) return;
      ref.invalidate(rebuildListProvider);
      ref.read(syncControllerProvider).nudge();
      setState(() => _loading = false);
      final router = GoRouter.of(context);
      router.go('/');
      router.push('/rebuild/$id');
    } catch (e) {
      if (mounted) {
        setState(() => _loading = false);
        ScaffoldMessenger.of(context).showSnackBar(
          SnackBar(content: Text(context.l10n.couldntAddSet('$e'))),
        );
      }
    }
  }

  @override
  Widget build(BuildContext context) {
    return AppButton(
      context.l10n.startSorting,
      variant: AppButtonVariant.hero,
      icon: Icons.playlist_add_check_rounded,
      expand: true,
      loading: _loading,
      onPressed: _loading ? null : _start,
    );
  }
}
